### PR TITLE
Re-enable use of tf message filter

### PR DESCRIPTION
**Pull request by shiveshkhaitan**
Re enable use of TF Filter in rviz_common and rviz_default_plugins. This connects to [issue#55](https://github.com/ros2/rviz/issues/55) [issue#359](https://github.com/ros2/rviz/issues/359)

### CODE REVIEW

**Comment by andreasholzner** [comment]
If I understand this correctly enabling tf filters does not make sure that tf is used as a transformer plugin via the `TransformerGuard` mechanism. This will likely fail if another transformer plugin were active.

On the other hand the ideas behind #346 were never realized since in fact tf is probably always used. Other usages of rviz without tf and ros messages will make use of completely independent display classes. So maybe it is best to revert #346 as this additional complexity seems unnecessary now and this would also remove the pointer casts from this pr. However, I am not sure how easy this can be done now -> @wjwwood .

**Comment by shiveshkhaitan** [comment]
Yes, I guess this defeats the whole purpose of #346. I will try incorporating tf filters into the plugin mechanism itself if it possible.

**Comment by Martin-Idel** [comment]
This will be very hard, I believe:

- The transformers are plugins, so which plugins exist is not known at compile time nor is it known which transformer is selected.
- The displays are likewise plugins and which plugins exist is not known at compile time. Since it depends on the message and generating custom messages is an important feature of ROS, there are likely many custom displays out there. 
- Message filters however require working with a subscription, which is in the `RosTopicDisplay`. Since the subscription is templated, it needs to be known at compile time. 

That's a problem, where I see only two possible solutions:

- Find a way to do an untemplated subscription, which will require changes to other libraries or at the very least a lot of new code (see how rosbag2 does it for normal ROS subscriptions)
- Treat the tf2 transformer as a special case: Introduce some code switching between a normal subscription and a message filter subscription based on whether the transformer is a tf2 transformer or a different transformer and declare that transformer plugins will only get a normal subscription.

I think the second solution is rather awkward and the first is quite hard.

Given that, I also believe it would be best to revert the introduction of the transformer plugin hierarchy, if @wjwwood agrees. If it's worthwhile, I could make a PR reverting it to see how difficult it is from a technical perspective.

**Comment by shiveshkhaitan** [comment]
Or can this be handled using something similar to ``transformer_guard``? 
Like the guard can check for the transformer_plugin been used. And if the plugin used is the default ``tf2 plugin``, subscription would be done using the ``tf_message_filters``. When using any other transformer, the normal subscription would be used. 
This way the current implementation of ``transformer_plugins`` would not be affected.

**Comment by Martin-Idel** [comment]
Yes, that's essentially what I tried to convey with my second option. That will be doable, I just wonder how understandable it'll be afterwards.

**Comment by shiveshkhaitan** [comment]
@Martin-Idel  @wjwwood  Is there any update on the mechanism to be followed?
Should I try implementing the  ```transform_guard``` approach and see how does it come up?

**Comment by wjwwood** [comment]
Sorry guys, this takes more thought and time than I have at the moment. I'll get to it as soon as I can.

**Comment by Myzhar** [comment]
Any news about this PR? Are message_filters enabled in Rviz2?

File: rviz_default_plugins/include/rviz_default_plugins/displays/laser_scan/laser_scan_display.hpp [hunk 1]
```diff
@@ -62,7 +63,7 @@ namespace displays
 // TODO(botteroa-si): This display originally extended the MessageFilterDisplay. Revisit when
```
Comment (jacobperron) [inline rviz_default_plugins/include/rviz_default_plugins/displays/laser_scan/laser_scan_display.hpp]: I think this TODO is resolved by this PR and can be removed.

File: rviz_default_plugins/src/rviz_default_plugins/displays/laser_scan/laser_scan_display.cpp [hunk 2]
```diff
@@ -66,13 +70,13 @@ void LaserScanDisplay::processMessage(sensor_msgs::msg::LaserScan::ConstSharedPt
 {
   // TODO(Martin-Idel-SI): Reenable once tf_filter is ported or delete if necessary
```
Comment (jacobperron) [inline rviz_default_plugins/src/rviz_default_plugins/displays/laser_scan/laser_scan_display.cpp]: This TODO is resolved and can be removed.

File: rviz_default_plugins/src/rviz_default_plugins/displays/map/map_display.cpp [hunk 3]
```diff
@@ -123,7 +123,7 @@ MapDisplay::MapDisplay()
     "Orientation", Ogre::Quaternion::IDENTITY, "Orientation of the map. (not editable)", this);
   orientation_property_->setReadOnly(true);
 
-  transform_timestamp_property_ = new rviz_common::properties::BoolProperty("Use Timestamp", false,
+  transform_timestamp_property_ = new rviz_common::properties::BoolProperty("Use Timestamp", true,
```
Comment (jacobperron) [inline rviz_default_plugins/src/rviz_default_plugins/displays/map/map_display.cpp]: Is this change necessary? I think we should leave the default value as-is.

File: rviz_default_plugins/include/rviz_default_plugins/displays/image/image_display.hpp [hunk 4]
```diff
@@ -41,14 +41,18 @@
...
 # include <OgreSharedPtr.h>
 
-# include "rviz_common/ros_topic_display.hpp"
+# include "rviz_common/message_filter_display.hpp"
 # include "rviz_common/render_panel.hpp"
 # include "rviz_common/properties/bool_property.hpp"
 # include "rviz_common/properties/float_property.hpp"
 # include "rviz_common/properties/int_property.hpp"
 # include "rviz_common/properties/queue_size_property.hpp"
+#include "rviz_common/transformation/frame_transformer.hpp"
```
Comment (jacobperron) [inline rviz_default_plugins/include/rviz_default_plugins/displays/image/image_display.hpp]: It looks like this include (and the newly included headers below) is not needed here. Please move those that are necessary to the cpp file instead.

Same goes for the similar changes to the other headers.

**Comment by jacobperron** [comment]
> Other usages of rviz without tf and ros messages will make use of completely independent display classes. 

Correct me if I'm wrong, but currently it looks like the only displays that require `tf2` are `TFDisplay`, `LaserScanDisplay`, and `RobotModelDisplay` (the ones making use of `TransformerGuard`). IIUC, the other displays can be used with other transformer plugins.

---

I spoke with @wjwwood, and we agree that it would be better keep the frame transformer abstraction.
Another solution is to update `tf2_ros::MessageFilter` to accept an interface that RViz frame transformers can implement. In this approach, we use the transformer plugin to get message transforms and leverage `tf2` for applying transforms. Updating `tf2_ros::MessageFilter` doesn't look like it will be straight-forward, but I can take a shot at it.

**Comment by jacobperron** [comment]
I've proposed changes to tf2 (https://github.com/ros2/geometry2/pull/121) and opened #422 as a possible approach for supporting the TF message filter with any FrameTransformer.

**Comment by AndreasAZiegler** [comment]
May I ask about the current state?

**Comment by jacobperron** [comment]
> May I ask about the current state?

ros2/geometry2#121 and #422 are ready for review. Once those are in, I suggest we rebase this PR to make use of the changes in order to remove the use of `tf2_ros::Buffer`.

**Comment by emersonknapp** [comment]
@shiveshkhaitan the above dependencies have been merged, are you still interested in following this through? I'm looking forward to this change!

**Comment by shiveshkhaitan** [comment]
Yes, I am currently working on the rebase. Should be getting it done in some time soon

File: rviz_common/include/rviz_common/message_filter_display.hpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+#include <tf2_ros/transform_listener.h>
+#include <tf2_ros/message_filter.h>
+
+#include <memory>
+
+#include "rviz_common/ros_topic_display.hpp"
+
+namespace rviz_common
+{
+/** @brief Display subclass using a rclcpp::subscription, templated on the ROS message type.
```
Comment (jacobperron) [inline rviz_common/include/rviz_common/message_filter_display.hpp]: nitpick: Replace use of `@brief ...` with `/// ...` (similar to other ROS 2 C++ code).

File: rviz_common/include/rviz_common/message_filter_display.hpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef RVIZ_COMMON__MESSAGE_FILTER_DISPLAY_HPP_
+#define RVIZ_COMMON__MESSAGE_FILTER_DISPLAY_HPP_
+
+#include <message_filters/subscriber.h>
+#include <tf2_ros/transform_broadcaster.h>
+#include <tf2_ros/transform_listener.h>
```
Comment (jacobperron) [inline rviz_common/include/rviz_common/message_filter_display.hpp]: I don't think we need to include headers for transform listener or broadcaster.

File: rviz_common/include/rviz_common/message_filter_display.hpp [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+ *
+ * This class handles subscribing and unsubscribing to a ROS node when the display is
+ * enabled or disabled. */
+template<class MessageType>
+class MessageFilterDisplay : public _RosTopicDisplay
+{
+// No Q_OBJECT macro here, moc does not support Q_OBJECT in a templated class.
+
+public:
+  /** @brief Convenience typedef so subclasses don't have to use
```
Comment (jacobperron) [inline rviz_common/include/rviz_common/message_filter_display.hpp]: nit: `@brief` -> `///`

File: rviz_common/include/rviz_common/message_filter_display.hpp [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+#include <memory>
+
+#include "rviz_common/ros_topic_display.hpp"
+
+namespace rviz_common
+{
+/** @brief Display subclass using a rclcpp::subscription, templated on the ROS message type.
+ *
+ * This class handles subscribing and unsubscribing to a ROS node when the display is
```
Comment (jacobperron) [inline rviz_common/include/rviz_common/message_filter_display.hpp]: The docs could be updated to clarify that this makes use of `message_filters` (not just an `rclcpp::subscription`).

File: rviz_common/include/rviz_common/message_filter_display.hpp [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+      tf_filter_->registerCallback(std::bind(&MessageFilterDisplay<MessageType>::incomingMessage,
+        this, std::placeholders::_1));
+      setStatus(properties::StatusProperty::Ok, "Topic", "OK");
+    } catch (rclcpp::exceptions::InvalidTopicNameError & e) {
+      setStatus(properties::StatusProperty::Error, "Topic",
+        QString("Error subscribing: ") + e.what());
+    }
+  }
+
+  void transformerChangedCallback() override
```
Comment (jacobperron) [inline rviz_common/include/rviz_common/message_filter_display.hpp:136]: The implementation of this method is identical to `updateTopic`. Maybe it's worth consolidating the logic in a common `resetSubscription()` method?

File: rviz_common/include/rviz_common/message_filter_display.hpp [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+  }
+
+  virtual void subscribe()
+  {
+    if (!isEnabled()) {
+      return;
+    }
+
+    if (topic_property_->isEmpty()) {
+      setStatus(properties::StatusProperty::Error,
```
Comment (jacobperron) [inline rviz_common/include/rviz_common/message_filter_display.hpp]: style nit: line break after the first open parenthesis

File: rviz_common/include/rviz_common/message_filter_display.hpp [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+
+    try {
+      subscription_ = std::make_shared<message_filters::Subscriber<MessageType>>(
+        rviz_ros_node_.lock()->get_raw_node(), topic_property_->getTopicStd(), qos_profile);
+      tf_filter_ =
+        std::make_shared<tf2_ros::MessageFilter<MessageType, transformation::FrameTransformer>>(
+        *context_->getFrameManager()->getTransformer(),
+        fixed_frame_.toStdString(), 10, rviz_ros_node_.lock()->get_raw_node());
+      tf_filter_->connectInput(*subscription_);
+      tf_filter_->registerCallback(std::bind(&MessageFilterDisplay<MessageType>::incomingMessage,
```
Comment (jacobperron) [inline rviz_common/include/rviz_common/message_filter_display.hpp]: nit: break after first open paren

File: rviz_common/include/rviz_common/message_filter_display.hpp [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+        this, std::placeholders::_1));
+      setStatus(properties::StatusProperty::Ok, "Topic", "OK");
+    } catch (rclcpp::exceptions::InvalidTopicNameError & e) {
+      setStatus(properties::StatusProperty::Error, "Topic",
```
Comment (jacobperron) [inline rviz_common/include/rviz_common/message_filter_display.hpp]: nit: break after first open paren

File: rviz_default_plugins/src/rviz_default_plugins/displays/pointcloud/point_cloud2_display.cpp [hunk 13]
```diff
@@ -179,13 +179,13 @@ void PointCloud2Display::update(float wall_dt, float ros_dt)
...
 {
-  RTDClass::reset();
+  MFDClass::reset();
   point_cloud_common_->reset();
 }
 
 void PointCloud2Display::onDisable()
 {
-  RosTopicDisplay::onDisable();
+  MessageFilterDisplay::onDisable();
```
Comment (jacobperron) [inline rviz_default_plugins/src/rviz_default_plugins/displays/pointcloud/point_cloud2_display.cpp]:
```suggestion
  MFDClass::onDisable();
```

Or, TBH, I think it would be more clear to update all instances of `MFDClass::` to `MessageFilterDisplay::`. Perhaps out of scope for this PR though.

File: rviz_default_plugins/src/rviz_default_plugins/displays/pointcloud/point_cloud_display.cpp [hunk 14]
```diff
@@ -68,13 +68,13 @@ void PointCloudDisplay::update(float wall_dt, float ros_dt)
...
 {
-  RTDClass::reset();
+  MFDClass::reset();
   point_cloud_common_->reset();
 }
 
 void PointCloudDisplay::onDisable()
 {
-  RosTopicDisplay::onDisable();
+  MessageFilterDisplay::onDisable();
```
Comment (jacobperron) [inline rviz_default_plugins/src/rviz_default_plugins/displays/pointcloud/point_cloud_display.cpp]:
```suggestion
  MFDClass::onDisable();
```

File: rviz_default_plugins/test/rviz_default_plugins/transformation/frame_transformer_tf_test.cpp [hunk 15]
```diff
@@ -45,8 +45,12 @@ class FrameTransformerTfFixture : public testing::Test
   FrameTransformerTfFixture()
   {
     auto clock = std::make_shared<rclcpp::Clock>(RCL_ROS_TIME);
+    if (!rclcpp::is_initialized()) {
+      rclcpp::init(0, nullptr);
```
Comment (jacobperron) [inline rviz_default_plugins/test/rviz_default_plugins/transformation/frame_transformer_tf_test.cpp]: If we call `rclcpp::init`, we should also call `rclcpp::shutdown` when we're finished.

I would add the static fixture methods:

```c++
  static void SetUpTestCase()
  {
    rclcpp::init(0, nullptr);
  }

  static void TearDownTestCase()
  {
    rclcpp::shutdown();
  }
```

**Review by jacobperron — COMMENTED** [review]
Thanks for doing this!

I've done a first pass. I haven't tested it yet.

File: rviz_default_plugins/include/rviz_default_plugins/displays/image/image_display.hpp (hunk 4 above)
Comment (jacobperron) [reply rviz_default_plugins/include/rviz_default_plugins/displays/image/image_display.hpp]: I think this comment is stale.

File: rviz_default_plugins/src/rviz_default_plugins/displays/map/map_display.cpp (hunk 3 above)
Comment (jacobperron) [reply rviz_default_plugins/src/rviz_default_plugins/displays/map/map_display.cpp]: Stale.

File: rviz_default_plugins/src/rviz_default_plugins/displays/pointcloud/point_cloud2_display.cpp (hunk 13 above)
Comment (shiveshkhaitan) [reply rviz_default_plugins/src/rviz_default_plugins/displays/pointcloud/point_cloud2_display.cpp]: I think we can do the conversion `MFDClass` -> `MessageFilterDisplay` in this PR itself. Should I do it?

File: rviz_common/include/rviz_common/message_filter_display.hpp (hunk 9 above)
Comment (bigdayangyu) [reply rviz_common/include/rviz_common/message_filter_display.hpp:136]: It looks like that `override` makes the compilation fail.
```
In file included from /home/lizo/ros_packaging_linux/ros2_overlay/src/rviz/rviz_default_plugins/include/rviz_default_plugins/displays/pointcloud/point_cloud_scalar_display.hpp:42:0,
                 from /home/lizo/ros_packaging_linux/ros2_overlay/src/rviz/rviz_default_plugins/include/rviz_default_plugins/displays/illuminance/illuminance_display.hpp:34,
                 from /home/lizo/ros_packaging_linux/ros2_overlay/src/rviz/rviz_default_plugins/src/rviz_default_plugins/displays/illuminance/illuminance_display.cpp:31:
/home/lizo/ros_packaging_linux/ros2_overlay/install/rviz_common/include/rviz_common/message_filter_display.hpp: In instantiation of ‘class rviz_common::MessageFilterDisplay<sensor_msgs::msg::Illuminance_<std::allocator<void> > >’:
/home/lizo/ros_packaging_linux/ros2_overlay/src/rviz/rviz_default_plugins/include/rviz_default_plugins/displays/pointcloud/point_cloud_scalar_display.hpp:65:35:   required from ‘class rviz_default_plugins::displays::PointCloudScalarDisplay<sensor_msgs::msg::Illuminance_<std::allocator<void> > >’
/home/lizo/ros_packaging_linux/ros2_overlay/src/rviz/rviz_default_plugins/include/rviz_default_plugins/displays/illuminance/illuminance_display.hpp:51:12:   required from here
/home/lizo/ros_packaging_linux/ros2_overlay/install/rviz_common/include/rviz_common/message_filter_display.hpp:134:8: error: ‘void rviz_common::MessageFilterDisplay<MessageType>::transformerChangedCallback() [with MessageType = sensor_msgs::msg::Illuminance_<std::allocator<void> >]’ marked ‘override’, but does not override
   void transformerChangedCallback() override
```

**Comment by bigdayangyu** [comment]
We rebased this PR again ontop of the latest nightly packaging job and with changing the `override` annotation it compiles successfully, however it segfaults when trying to launch RVIz.

File: rviz_common/include/rviz_common/message_filter_display.hpp (hunk 9 above)
Comment (shiveshkhaitan) [reply rviz_common/include/rviz_common/message_filter_display.hpp:136]: Wonder how does it build without any errors for me

File: rviz_default_plugins/src/rviz_default_plugins/displays/pointcloud/point_cloud2_display.cpp (hunk 13 above)
Comment (jacobperron) [reply rviz_default_plugins/src/rviz_default_plugins/displays/pointcloud/point_cloud2_display.cpp]: On second thought, I think we should keep using `MFDClass`. It looks like this was [added to avoid having to write the template type](https://github.com/ros2/rviz/blob/c05726637c6f91718d33e0bc9ee2717a688552b1/rviz_common/include/rviz_common/message_filter_display.hpp#L52-L54). It makes me wonder how this code is compiling to begin with.

File: rviz_common/include/rviz_common/message_filter_display.hpp [hunk 16]
```diff
@@ -41,18 +39,18 @@
 
 namespace rviz_common
 {
-/** @brief Display subclass using a rclcpp::subscription, templated on the ROS message type.
- *
- * This class handles subscribing and unsubscribing to a ROS node when the display is
- * enabled or disabled. */
+/// Display subclass using a rclcpp::subscription and tf2_ros::MessageFilter,
+/// templated on the ROS message type.
```
Comment (jacobperron) [inline rviz_common/include/rviz_common/message_filter_display.hpp]: Only the first line (ie the brief) should use `///`, the remaining docs can be put in `/** */`. [See our style guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#comments-and-doc-comments).

File: rviz_default_plugins/test/rviz_default_plugins/transformation/frame_transformer_tf_test.cpp (hunk 15 above)
Comment (jacobperron) [reply rviz_default_plugins/test/rviz_default_plugins/transformation/frame_transformer_tf_test.cpp]: Thanks for adding the shutdown, but I still think it would be better to add `rclcpp::init` to a SetUp method. Then we shouldn't need to check if `!rclcpp::is_initialized()`

File: rviz_default_plugins/src/rviz_default_plugins/displays/laser_scan/laser_scan_display.cpp [hunk 17]
```diff
@@ -50,29 +50,28 @@ LaserScanDisplay::LaserScanDisplay()
...
-//  {
-//    filter_tolerance_ = tolerance;
-//    tf_filter_->setTolerance(filter_tolerance_);
-//  }
-
+  rclcpp::Duration tolerance(static_cast<int32_t>(static_cast<rcl_duration_value_t>(
+      scan->time_increment * scan->ranges.size())), 0);
+  if (tolerance > filter_tolerance_) {
+    filter_tolerance_ = tolerance;
+    tf_filter_->setTolerance(filter_tolerance_);
```
Comment (jacobperron) [inline rviz_default_plugins/src/rviz_default_plugins/displays/laser_scan/laser_scan_display.cpp:73]: I'm having a problem rendering laser scans from Gazebo. I get the following message repeated:

```
[INFO] [rviz]: [signalFailure] Drop message: frame 'base_link' at time 2043.937 for reason(0)
[INFO] [rviz]: [signalFailure] Drop message: frame 'base_link' at time 2044.037 for reason(0)
[INFO] [rviz]: [signalFailure] Drop message: frame 'base_link' at time 2044.138 for reason(0)
...
```

Which is [coming from `tf2_ros::MessageFilter`](https://github.com/ros2/geometry2/blob/fccfd986ad82208dc64d0f1d650dc87cf336cae7/tf2_ros/include/tf2_ros/message_filter.h#L652). Specifically, I've traced the error message to [this line](https://github.com/ros2/geometry2/blob/fccfd986ad82208dc64d0f1d650dc87cf336cae7/tf2_ros/include/tf2_ros/message_filter.h#L539).

I probably introduced the bug in https://github.com/ros2/geometry2/pull/121. I think transform requests are timing out immediately due to a [zero timeout duration](https://github.com/ros2/geometry2/blob/fccfd986ad82208dc64d0f1d650dc87cf336cae7/tf2_ros/include/tf2_ros/message_filter.h#L327).

~~Aside from this bug, I'm not sure how the message filter worked properly in RViz before, since if the tolerance defaults to zero I wouldn't expect `processMessage` to be called at all. I'll continue digging and see if I can resolve the issue before moving forward with this PR.~~ Edit: This was a misunderstanding on my part about what the tolerance refers to.

File: rviz_common/include/rviz_common/message_filter_display.hpp (hunk 9 above)
Comment (jacobperron) [reply rviz_common/include/rviz_common/message_filter_display.hpp:136]: The `override` keyword makes sense to me, and it also builds for me. Perhaps you were building `rviz_default_plugins` against an older version of `rviz_common` somehow?

File: rviz_default_plugins/src/rviz_default_plugins/displays/laser_scan/laser_scan_display.cpp (hunk 17 above)
Comment (shiveshkhaitan) [reply rviz_default_plugins/src/rviz_default_plugins/displays/laser_scan/laser_scan_display.cpp:73]: Can you pls tell how to reproduce this. I was getting this error only when the fixed frame of rviz was not correct

Comment (jacobperron) [reply rviz_default_plugins/src/rviz_default_plugins/displays/laser_scan/laser_scan_display.cpp:73]: I tried running this [confbot_robot simulation tutorial](https://github.com/Karsten1987/confbot_robot/blob/master/tutorial/Linux_Tutorial.md#simulating-the-robot-in-gazebo) and when I change the fixed frame to the `odom` frame, I get the reported dropped messages and sensors stop rendering in RViz (laser and camera).

Comment (jacobperron) [reply rviz_default_plugins/src/rviz_default_plugins/displays/laser_scan/laser_scan_display.cpp:73]: I think this should fix the issue: https://github.com/ros2/geometry2/pull/146

**Comment by jacobperron** [comment]
Thanks @shiveshkhaitan! LGTM
I just rebased on the main branch so that CI will compile.

**Review by jacobperron — APPROVED** [review]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/7828/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3901/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6389/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/7665/)
